// @netlify/zip-it-and-ship-it/dist/archive.js
import { createWriteStream, readlinkSync } from 'fs';
import { promisify } from 'util';
import archiver from 'archiver';
import endOfStream from 'end-of-stream';
const pEndOfStream = promisify(endOfStream);
// Start zipping files
export const startZip = function (destPath) {
    const output = createWriteStream(destPath);
    const archive = archiver('zip');
    archive.pipe(output);
    return { archive, output };
};
// Add new file to zip
export const addZipFile = function (archive, file, name, stat) {
    if (stat.isSymbolicLink()) {
        const linkContent = readlinkSync(file);
        archive.symlink(name, linkContent, stat.mode);
    }
    else {
        archive.file(file, {
            name,
            mode: stat.mode,
            // Ensure sha256 stability regardless of mtime
            date: new Date(0),
            stats: stat,
        });
    }
};
// Add new file content to zip
export const addZipContent = function (archive, content, name) {
    archive.append(content, { name, date: new Date(0) });
};
// End zipping files
export const endZip = async function (archive, output) {
    archive.finalize();
    await pEndOfStream(output);
};
