// @netlify/edge-bundler/dist/node/config.js
import { promises as fs } from 'fs';
import { join } from 'path';
import { pathToFileURL } from 'url';
import tmp from 'tmp-promise';
import { getPackagePath } from './package_json.js';
// eslint-disable-next-line no-shadow
var ConfigExitCode;
(function (ConfigExitCode) {
    ConfigExitCode[ConfigExitCode["Success"] = 0] = "Success";
    ConfigExitCode[ConfigExitCode["UnhandledError"] = 1] = "UnhandledError";
    ConfigExitCode[ConfigExitCode["ImportError"] = 2] = "ImportError";
    ConfigExitCode[ConfigExitCode["NoConfig"] = 3] = "NoConfig";
    ConfigExitCode[ConfigExitCode["InvalidExport"] = 4] = "InvalidExport";
    ConfigExitCode[ConfigExitCode["RuntimeError"] = 5] = "RuntimeError";
    ConfigExitCode[ConfigExitCode["SerializationError"] = 6] = "SerializationError";
})(ConfigExitCode || (ConfigExitCode = {}));
const getConfigExtractor = () => {
    const packagePath = getPackagePath();
    const configExtractorPath = join(packagePath, 'deno', 'config.ts');
    return configExtractorPath;
};
export const getFunctionConfig = async (func, importMap, deno, log) => {
    // The extractor is a Deno script that will import the function and run its
    // `config` export, if one exists.
    const extractorPath = getConfigExtractor();
    // We need to collect the output of the config function, which should be a
    // JSON object. Rather than printing it to stdout, the extractor will write
    // it to a temporary file, which we then read in the Node side. This allows
    // the config function to write to stdout and stderr without that interfering
    // with the extractor.
    const collector = await tmp.file();
    // The extractor will use its exit code to signal different error scenarios,
    // based on the list of exit codes we send as an argument. We then capture
    // the exit code to know exactly what happened and guide people accordingly.
    const { exitCode, stderr, stdout } = await deno.run([
        'run',
        '--allow-env',
        '--allow-net',
        '--allow-read',
        `--allow-write=${collector.path}`,
        '--quiet',
        `--import-map=${importMap.toDataURL()}`,
        extractorPath,
        pathToFileURL(func.path).href,
        pathToFileURL(collector.path).href,
        JSON.stringify(ConfigExitCode),
    ], { rejectOnExitCode: false });
    if (exitCode !== ConfigExitCode.Success) {
        logConfigError(func, exitCode, stderr, log);
        return {};
    }
    if (stdout !== '') {
        log.user(stdout);
    }
    try {
        const collectorData = await fs.readFile(collector.path, 'utf8');
        return JSON.parse(collectorData);
    }
    catch {
        logConfigError(func, ConfigExitCode.UnhandledError, stderr, log);
        return {};
    }
    finally {
        await collector.cleanup();
    }
};
const logConfigError = (func, exitCode, stderr, log) => {
    switch (exitCode) {
        case ConfigExitCode.ImportError:
            log.user(`Could not load edge function at '${func.path}'`);
            log.user(stderr);
            break;
        case ConfigExitCode.NoConfig:
            log.system(`No in-source config found for edge function at '${func.path}'`);
            break;
        case ConfigExitCode.InvalidExport:
            log.user(`'config' export in edge function at '${func.path}' must be a function`);
            break;
        case ConfigExitCode.RuntimeError:
            log.user(`Error while running 'config' function in edge function at '${func.path}'`);
            log.user(stderr);
            break;
        case ConfigExitCode.SerializationError:
            log.user(`'config' function in edge function at '${func.path}' must return an object with primitive values only`);
            break;
        default:
            log.user(`Could not load configuration for edge function at '${func.path}'`);
            log.user(stderr);
    }
};
